Display type casting expressions on nodes

vitvakatu opened the pull request:
### Pull Request Description

The first part of #12312

Closes #12312

Type casting expressions can have incoming connections, detaching or reconnecting the connection changes the `expression` part (no touching `type annotation`). 
Supporting both valid ways of writing type annotations in self-argument position: 
- `expr:Type . method` — note the spacing for operator
- `(expr:Type).method`

First iteration (for history):

<img width="1043" alt="Screenshot 2025-04-07 at 12 52 06 PM" src="https://github.com/user-attachments/assets/88dfe7b4-3aeb-4544-b642-0c354216822b" />

Second iteration:

<img width="920" alt="Screenshot 2025-04-17 at 11 51 42 PM" src="https://github.com/user-attachments/assets/00fd73f7-c6a0-4639-8429-c0c67b60cd46" />


### Important Notes

In ~~self-argument~~ any position, only the `expression` part is editable, there is currently no way to change or remove `type annotation` without manual editing.
Also, the incoming self-argument arrow is not touching the icon directly.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

vitvakatu committed:
Add hiddenType field to ExpressionUpdate

vitvakatu committed:
Use SourceRange class instead of objects in aliasAnalysis tests

vitvakatu committed:
Displaying connections for type casts in nodes

vitvakatu committed:
Trying separate widget for type casting

vitvakatu committed:
Add TypeAnnotated AST node

vitvakatu committed:
Refactorings

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelfAccessChain.vue:
```diff
@@ -27,8 +27,18 @@ const { displayedIcon } = useDisplayedIcon(graph.db, toRef(tree, 'externalId'),
+  const input = WidgetInput.WithPort(WidgetInput.FromAstWithPortId(expression, portId))
+  const isTypeAnnotated = lhs instanceof Ast.Group || lhs instanceof Ast.TypeAnnotated
+  const showContents = isTypeAnnotated ? true : showFullAccessChain.value
+  input[DisplayIcon] = { icon: displayedIcon.value, showContents, noGap: isTypeAnnotated }
```
This breaks edges when parentheses are used in the self-access chain without a type annotation, e.g. `(file1).column_count`, because the `portId` points to the `Group`, but the `Group` isn't rendered because the `iconInput` is its `expression`.
I think the logic could be simplified by "unwrapping" parenthesized groups earlier.
This is getting complex enough I think it's worth unit testing.

vitvakatu replied:
A really nice catch, but I wouldn’t say this implementation “breaks” it — it wasn’t working in the first place on develop! It is now fixed and the code is simplified.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelfAccessChain.vue:
```diff
@@ -40,15 +50,31 @@ const showFullAccessChain = computed(() => /^[A-Z]/.test(props.input.value.lhs?.
 
 <script lang="ts">
+/** Subject of the self access chain can be either a regular expression, or a type annotated expression. */
+const extractSubject = (accessChain: Ast.PropertyAccess | undefined) => {
```
Why not a `function`?

vitvakatu replied:
Is there a substantial difference? Anyway, this thing is removed now.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelfAccessChain.vue:
```diff
@@ -40,15 +50,31 @@ const showFullAccessChain = computed(() => /^[A-Z]/.test(props.input.value.lhs?.
       const selfId = tree.potentialSelfArgumentId
-      const match = selfId != null && info.input.value.lhs?.id === selfId
-      return match ? Score.Good : Score.Mismatch
+      const subject = extractSubject(info.input.value)
```
If the `potentialSelfArgumentId` analysis could identify the `PropertyAccess` AST node instead, it would avoid adding reactive dependencies for every expression that is tested against this criterion.

vitvakatu replied:
Great idea, although it brings some complications in `node.ts`. Implemented it, avoiding this complex unwrapping here.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetTypeCast.vue:
```diff
@@ -0,0 +1,77 @@
+    score: (info) => {
+      // Only groups with TypeAnnotated expression are valid.
+      if (
+        info.input.value instanceof Ast.Group &&
```
Why does this widget need to be concerned with groups? Can that be left to `WidgetGroup`?

vitvakatu replied:
True, removed it.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetTypeCast.vue:
```diff
@@ -0,0 +1,77 @@
+
+<template>
+  <div class="WidgetTypeCast">
+    <NodeWidget v-if="!isSelfArgument" :input="expressionInput" />
```
I don't see why the widget hierarchy is inverted from the tree hierarchy in this case, with `WidgetIcon` being a parent of the `WidgetTypeCast` and the `WidgetTypeCast` displayed as if it didn't have an expression--if the `WidgetIcon` matched the annotation's `expression` and were a child of the `WidgetTypeCast`, I think it should be possible to avoid special handling for self arguments in `WidgetTypeCast`.

vitvakatu replied:
Perhaps I’m missing something, but I didn’t change this part for the following reasons:
1. I don’t see a particular inversion here. In fact, it is precisely how the `SelfAccessChain` widget works currently — the icon is a parent of the method invocation.
2. The current implementation only instantiates the icon in one place—the `SelfAccessChain` widget—instead of duplicating the instantiation and configuration in `WidgetTypeCast`.
3. We won’t avoid handling of self arguments in `WidgetTypeCast` anyway, because we still need to understand when we’re in the self-argument position — and when we should display the icon. So the check will still be there, just with a different action.

kazcw commented on app/gui/src/project-view/util/ast/node.ts:
```diff
@@ -70,14 +70,27 @@ export function primaryApplicationSubject(
-      subject instanceof Ast.Wildcard
-    )
-  )
-    return
+    subject instanceof Ast.Group &&
+    subject.expression instanceof Ast.TypeAnnotated &&
+    isAcceptableSubject(subject.expression.expression)
+  ) {
+    return {
+      subject: subject.expression.expression.id,
+      accessChain: accessChain.map((ast) => ast.id),
+    }
+  }
+  if (subject instanceof Ast.TypeAnnotated && isAcceptableSubject(subject.expression)) {
+    return {
+      subject: subject.expression.id,
+      accessChain: accessChain.map((ast) => ast.id),
+    }
+  }
+  // The leftmost element must be an identifier or a placeholder.
+  if (!isAcceptableSubject(subject)) return
```
I think this could be simplified by using some conditional unwrapping steps, and then the `isAcceptableSubject`/`accessChain` logic would only need to be in one place.
Something like
```ts
let innerExpression = expression
if (innerExpression instanceof Ast.Group && innerExpression.expression) innerExpression = innerExpression.expression
if (innerExpression instanceof Ast.TypeAnnotated) innerExpression = innerExpression.expression
// innerExpression can be handled uniformly
```

vitvakatu replied:
Ouch, I actually missed this part of code during refactoring, thanks for spotting!

vitvakatu replied:
Simplified this code greatly.

kazcw reviewed (commented):
As this is a complex feature for users, I think it's important for it to behave as consistently as possible--why render the whole `TypeAnnotated` as a connected port in self-argument position, but only the inner `expression` as a port in other positions?

vitvakatu committed:
Crude implementation of displaying methods from hidden types in CB

vitvakatu committed:
Showing multiple types in CB top bar

vitvakatu committed:
Better filtering

vitvakatu committed:
applying type casting suggestions

vitvakatu committed:
Tests, refactorings, and bug fixes

vitvakatu committed:
Addressing review comments

vitvakatu committed:
Merge with develop

vitvakatu commented:
> As this is a complex feature for users, I think it's important for it to behave as consistently as possible--why render the whole TypeAnnotated as a connected port in self-argument position, but only the inner expression as a port in other positions?

Fair point, the simple answer is that it was far easy to implement. Updated version of the code shows both uniformly, with the whole thing being a port (but only the inner part is editable in graph editor).

kazcw commented on app/gui/src/project-view/stores/graph/graphDatabase.ts:
```diff
@@ -652,3 +658,26 @@ export type Node = NodeDataFromAst &
+  a: PrimaryApplication | undefined,
+  b: PrimaryApplication | undefined,
+) {
+  if (a == null || b == null) return false
```
`primaryApplicationEquals(undefined, undefined) === false`?

vitvakatu replied:
Fixed!

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetPort.vue:
```diff
@@ -143,6 +143,13 @@ export const widgetDefinition = defineWidget(
+      // in argument positions. A port needs to be created with portId override, and it
+      // will be handled by `WidgetTypeCastPort`. Without this check, the port will be
+      // created with invalid portId because of `ArgumentInfoKey` being set on the input.
+      if (props.input.portId.includes('TypeAnnotated')) {
```
I prefer to avoid inspecting an AstId--they contain the type name to aid debugging, but they are intended to be opaque values. In the future we may change the format--the implementation of some of the #10752 changes could be much  more efficient if we use plain numbers for AstIds.
Maybe here we could check if the input is an `Ast`, and then if the `portId` doesn't match the `AstId` we don't create a port unless `forcePort` is set?

vitvakatu replied:
Yeah, it’s definitely an ugly hack, but I’m struggling to find a solution. The issue here is that `portId` **equals** `AstId` at this moment, we change it only in `WidgetTypeCast`, which has lower priority, so it matches later. 
I’m not sure there is some order of priorities in which it starts to work without breaking some cases. The only way to know that we don’t want to create a port immediately is to inspect the AST and see that it is actually a TypeAnnotated, but I don’t want to do that.
It looks to me that we need some kind of bottom-up propagation of information through the tree to implement it correctly.

Although I found a better check, without comparing with the string:
```
props.input.value instanceof Ast.TypeAnnotated && props.input.value.id === props.input.portId
```

kazcw commented on app/gui/src/project-view/stores/graph/graphDatabase.ts:
```diff
@@ -369,7 +369,13 @@ export class GraphDb {
       const astFields: NodeAstField[] = ['outerAst', 'pattern', 'rootExpr', 'innerExpr']
       astFields.forEach(updateAst)
-      if (oldNode.primarySubject !== primarySubject) node.primarySubject = primarySubject
+      if (
+        !primaryApplicationEquals(
+          oldNode.primaryApplication as PrimaryApplication,
+          primaryApplication,
+        )
+      )
+        node.primaryApplication = primaryApplication
```
I think in this approach there's some hazard of future changes causing reactivity bugs. Elsewhere we read the `primaryApplication` field and pass around the resulting object, which has no direct reactive connection to future updates if we replace the object wholesale. It would be more robust to conditionally update each field of the `PrimaryApplication`.

vitvakatu replied:
Implemented as suggested, but the code is… not nice. Can I improve anything? https://github.com/enso-org/enso/pull/12751/commits/65acd4276160c610b239c95024f256bdc47aeecd

kazcw replied:
To ensure it's deeply reactive, the object would need to never be replaced with a different object--this includes if the value goes to `undefined` and then is set to a new value. So instead of the object being nullable, its fields must be nullable. If the object is always present and its fields are nullable, the update can follow the `if (oldNode.primarySubject !== primarySubject) node.primarySubject = primarySubject` pattern for each field.

vitvakatu replied:
Done

vitvakatu committed:
Update code in graphDatabase.test.ts to use SourceRange class instead of
objects

vitvakatu committed:
Conditionally update primaryApplication in node info

vitvakatu committed:
Avoid using string comparison for portId

vitvakatu committed:
Use template with v-if for repeated components

vitvakatu commented:
Component browser changes from #12763 rebased and added on top. See this commit range if you want easier review: https://github.com/enso-org/enso/pull/12751/files/41725162736380435333e1c5cdfe46e21b64faf9..c85371602070e267f6ba8ec9a607b635af89e705

<blockquote><img src="https://avatars.githubusercontent.com/u/6566674?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/pull/12751">Display type casting expressions on nodes by vitvakatu · Pull Request #12751 · enso-org/enso</a></strong></div><div>Pull Request Description
The first part of #12312
Closes #12312
Type casting expressions can have incoming connections, detaching or reconnecting the connection changes the expression part (no touc...</div></blockquote>

vitvakatu committed:
Update changelog

vitvakatu committed:
Fix CB tests

vitvakatu committed:
Make fields of PrimaryApplication nullable, check each field
individually when reactively updating the node in graphDatabase

vitvakatu committed:
Make clippy happy

kazcw commented on app/rust-ffi/src/lib.rs:
```diff
@@ -1,5 +1,6 @@
 use wasm_bindgen::prelude::*;
 
+use enso_parser::lexer::test::TokenOperatorProperties;
```
That's a test reexport, the proper path is something like `enso_parser::syntax::token::TokenOperatorProperties`

vitvakatu replied:
Ah, auto-imports in IDE…

kazcw commented on app/gui/src/project-view/stores/project/computedValueRegistry.ts:
```diff
@@ -79,13 +81,15 @@ function updateInfo(
 ) {
   const newInfo = combineInfo(info, update, projectNames)
   if (newInfo.typename !== info.typename) info.typename = newInfo.typename
+  if (newInfo.hiddenTypes !== info.hiddenTypes) info.hiddenTypes = newInfo.hiddenTypes
```
`combineInfo` always creates a new `hiddenTypes` array. Need a deep comparison like `arrayEquals` with a `.equals` predicate.

vitvakatu replied:
Nice catch, I wish we had a lint for this

vitvakatu committed:
Fix import in rust-ffi

vitvakatu committed:
Use arrayEquals for hiddenTypes in computedValueRegistry

vitvakatu committed:
Merge develop

farmaazon commented on app/gui/src/project-view/stores/graph/graphDatabase.ts:
```diff
@@ -652,3 +669,21 @@ export type Node = NodeDataFromAst &
+
+export interface PrimaryApplication {
+  /** A child AST in a syntactic position to be a self-argument input to the node. */
+  potentialSelfArgument: Ast.AstId | null
```
Why is it "potential"? I guess because sometimes this is a type? I would add this clarification.

vitvakatu replied:
Fair point, I was reusing previous naming. It really can be only Ident, Wildcard or null, so I renamed it to just `selfArgument`. I’m not sure if we can have an Ident in this position which is not self-argument, I hope not.

farmaazon commented on app/gui/src/project-view/util/ast/node.ts:
```diff
@@ -44,41 +45,54 @@ export function nodeFromAst(ast: Ast.Statement, isOutput: boolean): NodeDataFrom
+    function: null,
+    accessChain: null,
+    potentialSelfArgument: null,
+  }
```
Let's make this `emptyPrimaryApplication` a function, and use it also in line 37

vitvakatu replied:
Done

farmaazon commented on app/gui/src/project-view/util/ast/node.ts:
```diff
@@ -44,41 +45,54 @@ export function nodeFromAst(ast: Ast.Statement, isOutput: boolean): NodeDataFrom
+  }
+  if (subject instanceof Ast.TypeAnnotated && subject.expression) {
+    subject = subject.expression
+  }
```
At least for groups, I think this should be a loop.

Add a test case with `((selfArg)).foo`

vitvakatu replied:
Done. Here, I also need to go into `TypeAnnotated,` so using `unwrapGroups` is impossible.

farmaazon commented on app/gui/src/project-view/util/ast/__tests__/aliasAnalysis.test.ts:
```diff
@@ -338,3 +339,10 @@ test(
     «3,default» = 1
     »3,default«`),
 )
+
+test(
+  'Type casting',
+  runTestCase(`«1,main» =
+    «2,x» = 1
+    (»2,x«:Int).toString`),
+)
```
what about case:
```
   x = 1
   x : Int
   x.toString
```
We should not have a connection to the second line. But I'm not sure if such "offline type annotations" are still supported.

vitvakatu replied:
This syntax seems unsupported by the language; at least, it does not compile. This whole test is also kind of unnecessary because alias analysis will work as long as we have an identifier somewhere in the tree, and TypeAnnotated is not difficult in the sense that it does not require any additional logic, unlike closures and functions.

farmaazon commented on app/gui/src/project-view/providers/widgetRegistry.ts:
```diff
@@ -20,6 +20,17 @@ export namespace WidgetInput {
     }
   }
 
+  /** Returns widget-input data for the given AST tree or token with a specific port ID. */
+  export function FromAstWithPortId<A extends Ast.Ast | Ast.Token>(
+    ast: A,
+    portId: PortId,
+  ): WidgetInput & { value: A } {
+    return {
+      portId,
+      value: ast,
+    }
+  }
```
This should be used inside `FromAst` - fewer complex type annotations.

vitvakatu replied:
Done

farmaazon commented on app/gui/src/project-view/components/FunctionSignatureEditor.vue:
```diff
@@ -77,13 +77,17 @@ const rootStyle = computed(() => {
 })
+
+// We surely don’t have primary application for the function definition.
+const primaryApplication = { function: null, accessChain: null, potentialSelfArgument: null }
```
anther usage of `emptyPrimaryApplication` function.

Or we should consider just making this field nullable? It shouldn't be bad, we could just add `?` to most usages.

vitvakatu replied:
Done. It can’t be nullable for reactivity reasons, as pointed out by Kaz in this thread: https://github.com/enso-org/enso/pull/12751#discussion_r2050377740

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetTypeCast.vue:
```diff
@@ -0,0 +1,60 @@
+
+<script lang="ts">
+export const widgetDefinition = defineWidget(
+  IsTypeCastKey,
```
We don't check if the value is actually `true`

If we don't care, because it's always true, then the field's type should also be `true` (instead of `boolean`)

vitvakatu replied:
Done

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelfAccessChain.vue:
```diff
@@ -27,8 +27,22 @@ const { displayedIcon } = useDisplayedIcon(graph.db, toRef(tree, 'externalId'),
-  input[DisplayIcon] = { icon: displayedIcon.value, showContents: showFullAccessChain.value }
+  let expression = lhs
+  let portId = lhs.id
+  // Unwrap the Group if present.
+  if (expression instanceof Ast.Group && expression.expression) {
+    expression = expression.expression
+    portId = expression.id
+  }
+  if (expression instanceof Ast.TypeAnnotated) {
```
Again, there might be multiple groups. And annotations perhaps too?

vitvakatu replied:
Here I’m using `unwrapGroups` now. I tried looking into nested type annotations, but the engine does not support them, as they seem to be recognized as “type operator chains”. I feel this implementation is robust against the most obvious syntactic variations that the user can reasonably come up with.

vitvakatu replied:
And of course, it is not clear how nested type annotations should translate into a self-access chain anyway, so let’s leave it to the future selves.

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetPort.vue:
```diff
@@ -143,6 +143,16 @@ export const widgetDefinition = defineWidget(
     score: (props, _db) => {
       const portInfo = injectPortInfo(true)
       const value = props.input.value
+      // This is a workaround to avoid automatic port around type annotated expressions
+      // in argument positions. A port needs to be created with portId override, and it
+      // will be handled by `WidgetTypeCastPort`. Without this check, the port will be
+      // created with invalid portId because of `ArgumentInfoKey` being set on the input.
+      if (
+        props.input.value instanceof Ast.TypeAnnotated &&
+        props.input.value.id === props.input.portId
+      ) {
+        return Score.Mismatch
+      }
```
`props.input.value` is aliased just a line above.

Also, shouldn't this kind of problems be resolved by `WidgetTypeCastPort` having higher priority?

vitvakatu replied:
Thanks for noting. 
> Also, shouldn't this kind of problems be resolved by WidgetTypeCastPort having higher priority?

It's not that easy because the priority requirements are quite nuts here. For example, `WidgetTypeCastPort` must match after `Group`, but `Group` has priority 999, and `Group` must have a lower priority than the `WidgetPort` in order not to break some common cases unrelated to type casting. So we have a strict requirement `Port < Group < WidgetTypeCastPort`, while we need `WidgetTypeCastPort < Port` to avoid this hack.

And I’m sure there are about a dozen similar unsolvable equations with other widgets; this one is the most obvious.

farmaazon commented on app/gui/src/project-view/components/ComponentBrowser/input.ts:
```diff
@@ -134,8 +135,15 @@ export function useComponentBrowserInput(
+    const info = graphDb.getExpressionInfo(definition)
+    if (info == null) return null
+    const typename = info.typename
+    const additionalTypes = info.hiddenTypes
```
```suggestion
    const { typename, additionalTypes } = info
```

vitvakatu replied:
Done

farmaazon commented on app/gui/src/project-view/components/ComponentBrowser/input.ts:
```diff
@@ -134,8 +135,15 @@ export function useComponentBrowserInput(
+    const additionalTypes = info.hiddenTypes
+    if (typename != null) {
+      const entry = suggestionDb.getEntryByProjectPath(typename)
+      if (entry) additionalTypes.push(...suggestionDb.ancestors(entry))
```
I think the ancestors are bit different thing; they do not need any typecasting (unless there's name conflict... but in that case library author might have a reason to overshadow).

vitvakatu replied:
Good catch. It caused type casting to be inserted when it was unnecessary. Now, ancestors are separate things and are handled slightly differently.

farmaazon commented on app/gui/src/project-view/components/ComponentBrowser/input.ts:
```diff
@@ -168,7 +176,20 @@ export function useComponentBrowserInput(
     requiredImport: ProjectPath | undefined
   } {
-    if (sourceNodeIdentifier.value) {
+    const displayOwner = (owner: ProjectPath) => {
+      if (owner.path) return qnLastSegment(owner.path)
+      if (owner.project) return qnLastSegment(owner.project)
+      return 'Main' as Ast.Identifier
+    }
```
Is it not a method of `ProjectPath`? If not, I still think it could be. It doesn't seem to be only about displaying - in the code we want to put `Main`s as well.

vitvakatu replied:
Added some helpers and refactored this code.

farmaazon commented on app/gui/src/project-view/components/ComponentBrowser/input.ts:
```diff
@@ -168,7 +176,20 @@ export function useComponentBrowserInput(
+    }
+    if (sourceNodeIdentifier.value && sourceNodeType.value?.type === 'known') {
+      const sourceType = sourceNodeType.value.typename
+      const owner = entry.kind === SuggestionKind.Method ? entry.memberOf : undefined
```
Constructors also have owners.

vitvakatu replied:
Good catch, fixed.

farmaazon commented on app/gui/src/project-view/components/ComponentBrowser/filtering.ts:
```diff
@@ -280,13 +301,19 @@ export class Filtering {
 
   /** TODO: Add docs */
-  filter(entry: SuggestionEntry, additionalSelfTypes: ProjectPath[]): MatchResult | null {
+  filter(entry: SuggestionEntry): MatchResult | null {
```
On this occasion, add docs to this function.

vitvakatu replied:
Added.

farmaazon commented on app/gui/integration-test/project-view/componentBrowser.spec.ts:
```diff
@@ -100,14 +100,14 @@ test('Opening Component Browser from output port buttons', async ({ page }) => {
   await expect(createNodeFromPortButton).toBeVisible()
   await createNodeFromPortButton.click({ force: true })
-  await expectAndCancelBrowser(page, '', null, 'selected')
+  await expectAndCancelBrowser(page, '', 'Input Components', 'selected')
```
The tests were right: we don't want to display `'Input Components'` with source node.

vitvakatu replied:
Thanks. I was a bit confused by the fact that tests differ from the actual IDE behavior. The reason is that in tests, we don’t have expression info available on some of the nodes, so no label is indeed visible. In practice, though, we have type information for all nodes, so the label is visible almost always.

vitvakatu committed:
Addressing review comments

- Revert changes to integration tests, the bug in input.ts module fixed
- Separate type ancestors from additional types, avoiding type cast when
  using ancestor methods
- Fix issue with modifying expression info database when filtering
- Add docs for `filter` function
- Refactor `inputAfterApplyingSuggestion` function to reuse code from
  ProjectPath implementation, adding some additional helpers there
- `emptyPrimaryApplication` helper
- Remove unused `showFullAccessChain` from SelfAccessChain widget
- Use `unwrapGroups` function copied from another PR
- Better logic for unwrapping nested groups and type annotations in
  node.ts and SelfAccessChain code. Nested type annotations are
  apparently not supported by the engine.
- Rename `potentialSelfArgument` to `selfArgument`

vitvakatu committed:
Merge develop

vitvakatu committed:
Update changelog after merge

vitvakatu committed:
Merge develop
